vllm/distributed/communication_op.py

```diff
@@ -5,10 +5,41 @@
 
 from .parallel_state import get_tp_group
 
+torch.library.define("vllm::tp_out_of_place_ar",
+                     ("(Tensor(a!) input_ ) -> Tensor"))
 
-def tensor_model_parallel_all_reduce(input_: torch.Tensor) -> torch.Tensor:
+
+@torch.library.register_kernel("vllm::tp_out_of_place_ar", ("cuda", "cpu"))
```

Comment:
These can probably be combined to use `torch.library.custom_op`, e.g.

```
@torch.library.custom_op(mutates_args=["input_"], device_types=("cuda", "cpu"))
def _tp_out_of_place_ar ...
```

similarly for the rest.